### PR TITLE
Start a roscore for a job

### PR DESCRIPTION
**Pull request by eggerk**
implements #86

### CODE REVIEW

File: run_build_impl.sh [hunk 1]
```diff
@@ -286,6 +291,41 @@ cd $WORKSPACE
 # Prepare cppcheck ignore list. We want to skip dependencies.
 CPPCHECK_PARAMS="src --xml --enable=missingInclude,performance,style,portability,information -j8 -ibuild -i$DEPS"
 
+function kill_roscore_on_exit {
+  if $START_ROSCORE && $ROS_PID -gt 0 ; then
```
Comment (HannesSommer) [inline run_build_impl.sh]: I think you are missing a `[[  ..  ]] ` around `$ROS_PID -gt 0` . And I'd suggest a `""` around `$ROS_PID` to be robust against an empty value for `ROS_PID`

File: run_build_impl.sh [hunk 2]
```diff
@@ -286,6 +291,41 @@ cd $WORKSPACE
...
+  fi
+
+  # Start roscore.
+  export ROS_MASTER_URI="http://localhost:$ROS_PORT"
+  echo "Starting roscore on port $ROS_PORT."
+  roscore -p $PORT > /dev/null & 
+  ROS_PID=$!
+  trap kill_roscore_on_exit EXIT
+fi
+
```
Comment (HannesSommer) [inline run_build_impl.sh]: Well done!

**Comment by eggerk** [comment]
test this please

**Comment by HannesSommer** [comment]
The slave-master is a config bug and  slaves-mc-xenial-kinetic is missing the apt-get install permissions. So both can be ignored.
So apparently it does not break anything - if not used. Now we need to use it. 
I suggest adding a test package to this repository, like https://github.com/ethz-asl/continuous_integration/tree/master/ci_test_pkg.
Maybe just a python test that connects to the roscore and is green on success.

File: run_build_impl.sh [hunk 3]
```diff
@@ -286,6 +291,41 @@ cd $WORKSPACE
...
+  if [ $ROS_PORT -lt 0 ] ; then
+    echo "Couldn't find an unused port for the roscore." >&2
+    exit 1
+  fi
+
+  # Start roscore.
+  export ROS_MASTER_URI="http://localhost:$ROS_PORT"
+  echo "Starting roscore on port $ROS_PORT."
+  roscore -p $PORT > /dev/null &
```
Comment (HannesSommer) [inline run_build_impl.sh]: Here it should be `$ROS_PORT`. That is why it currently fails.

**Comment by eggerk** [comment]
It works now when there is a roscore. Otherwise the new unit test fails.

There is also the problem that the slaves don't have `lsof` installed. Now it can't check for free ports and just uses the first one in the list.

**Comment by HannesSommer** [comment]
I've installed lsof on the slaves. Maybe you should make a explicit check before whether the command lsof exists. E.g. with `command -v lsof >/dev/null`. And fail if it does not exist.

**Comment by HannesSommer** [comment]
Ah, oh, you did that already! Awesome.

**Comment by HannesSommer** [comment]
Btw. I've learned that `command -v` should be more reliable, when using bash, than which, because `which` is less standardized.

**Comment by eggerk** [comment]
Ok, I can quickly change that

**Comment by HannesSommer** [comment]
I had missed the mc-xenial-kinetic slave...

**Comment by HannesSommer** [comment]
This should be it. Very well done @eggerk !
Feel free to merge!

Minor remark:
There is a minor conceptual problem concerning the shared resources ports: imagine two tests start at the same time on one node: they might still try the same port! The usual way to avoid that would be to hold an exclusive lock while scanning for free ports. Problem is, we can control this script but other processes might grab that port as well. With those we cannot agree on a locking mechanism.
The real good pragmatic way here would therefore probably be to try running the roscore and if it fails because it cannot listen on the port then start over again with searching for a free port. 
But that would be overkill here. After all the risk is very low because probability is low since it is quite a short time that it takes to search for a free port compared to the average period with which this will happen.

**Comment by pleemann** [comment]
Just out of curiosity (and assuming that the roscore is required for rostests only?): Is it necessary to start the roscore by hand? According to [rostest documentation](http://wiki.ros.org/rostest/Connecting%20to%20the%20rostest%20master): "By default rostest uses a random port so many rostests can be run in parallel."

**Comment by HannesSommer** [comment]
Thanks @pleemann , for this question! 
I wasn't aware of this possibility. My only contact to jobs requiring ROS is through our Jenkins setup. So far on each node people kind of manually started one `roscore` for all the jobs together. Of course this caused trouble and we didn't want to copy that concept to our new Docker based Jenkins slaves. So we came up with this solution. 
Now I would assume that for some reason `roscore` is / was required somewhere without `rostest`. 
@eggerk,  could mention this alternative way to get a `roscore` in the wiki? Maybe point to this (http://docs.ros.org/kinetic/api/catkin/html/howto/format2/rostest_configuration.html).

**Pull request merged**